### ISSUE TITLE
Add Filter Buttons for Disconnected Services in the Applications View

### ISSUE THREAD

**Issue by apriofrost** [chef/automate#1780]
## Description

Allow users to find and remediate disconnected nodes by adding filter buttons for disconnected services for both the service group table and the service detail list(right side-panel).

## Definition of Done
![image.png](https://images.zenhubusercontent.com/5a721a104b5806bc2bc51768/210957fb-3d68-45dd-9b41-d030159721e6)
* Add a big filter button for disconnected services groups above the service groups table
* Add a small filter button for disconnected services above the service details list
* For the big filter, when it's selected, only service groups that have one or more disconnected services are shown in the service groups table
* For the small filter, when it's selected, only disconnected services are shown in the services detail list.
* Responsive filter size
  - For big filters: when resizing the screen, they respond to the new size while keeping the same padding and a consistent `8px` margin from each other. The total width of the big filters should match with that of the table underneath. The minimal width of the individual big filter is `115px`. Create a new line when the existing line is full with all filters at the minimum width.
<img src="https://images.zenhubusercontent.com/5a721a104b5806bc2bc51768/0c5bc69e-d98e-4cdd-8161-2fbec3712104" width="200">
  - For small filters: currently, the width of the side-bar is fix-width. The width of the smaller filters only responds to its content (the count numbers) while keeping the consistent padding and margin; create a new line when the existing one is full. 

### Backend 
Done, see [PR](https://github.com/chef/automate/pull/1641)

* API for service groups can filter for disconnected
* API for health counts includes the count of disconnected
* API for services by service group can filter for disconnected
* API for health counts for services in a service group can filter for disconnected

### UI Demo
View demo video in Google Drive: https://drive.google.com/open?id=1YMnYQq7ViS4yhgijVZqhtAREfMc2h342
#### Implementation notes
* Reach out to Liz to invite you to the InVision project so that you could use the inspect feature to see the styling details.
* Styles for both filters:
  * Reuse the existing `icon/disconnected service` status icon
  * when hovered, the icon color changes from `#C22DD5` to `$chef-info-alert`; the background color changes from `FFFFFF` to `$chef-info-alert-light`; the background border becomes hidden.
  * when selected, add the background border back to 1px, inside, `#C22DD5`.
* Mockup link: https://chef.invisionapp.com/share/C3T08NGNKEQ#/388380424_step_2-_add_single_select_filter

**Comment by scottopherson** [comment]
@chef/ui-team @chef/superteam jotted down a couple notes and questions for no one in particular:

The service-groups API and UI are starting to have some weird ergonomics between the two

![](https://user-images.githubusercontent.com/479121/66874772-7e3c5b00-ef69-11e9-94ca-ea274961d404.png)

Currently, these filter buttons above the service-group list are mapped to the `status` key within the `filter` query param in `/service-groups`:

![](https://user-images.githubusercontent.com/479121/66874808-9b712980-ef69-11e9-80cb-5c6cb1c177fb.png)

```
/apis/beta/applications/service-groups
  ?filter=status:warning
  &sorting.field=percent_ok
  &sorting.order=ASC
  &pagination.page=1
  &pagination.size=25
```

And currently, these filter buttons above the services list in the service-group sidebar are mapped to the `health` query param in `/service-groups/:group_id`:

![](https://user-images.githubusercontent.com/479121/66874787-8a281d00-ef69-11e9-8e1d-c8517296a67e.png)
```
/apis/beta/applications/service-groups/:group_id
  ?health=ok
  &pagination.page=1
  &pagination.size=25
```

Now we're wanting to be able to filter on `disconnected` service groups and services.

![](https://user-images.githubusercontent.com/479121/66875066-441f8900-ef6a-11e9-92fd-7971a54fa8f5.png)

For `/service-groups/:group_id`, https://github.com/chef/automate/pull/1641 added the ability to filter on `disconnected` services by adding a `disconnectedStatus` key within the `filter` query param:

```
/apis/beta/applications/service-groups/:group_id
  ?filter=disconnectedStatus:disconnected
  &pagination.page=1
  &pagination.size=25
```

Did we add any `disconnected` filtering for `/service-groups` endpoint as well?

So we now have:

```
/service-groups/:group_id

health=critical
health=warning
health=ok
health=unknown
filter=disconnectedStatus:disconnected

/service-groups

filter=status:critical
filter=status:warning
filter=status:ok
filter=status:unknown
filter=disconnectedStatus:disconnected (maybe? Not sure yet if `/service-groups` supports filtering on `disconnected`)
```

Would it be possible to simplify things to something like:

```
filter=status:critical
filter=status:warning
filter=status:ok
filter=status:unknown
filter=status:disconnected
```

for both `/service-groups` and `/service-groups/:group_id`?

Not a blocker for implementing this feature but would make it a lot easier to not have a combination of `health`, `filter=status`, and `filter=disconnectedStatus` API params, particularly since the state of these UI filters are serialized to the browser URL params as well.

If there's reasons where we actually do need to have an separate `status` type for `disconnected`:

```
filter=status:critical
filter=status:warning
filter=status:ok
filter=status:unknown
filter=disconnectedStatus:disconnected
```

does it still make sense to present `status` and `disconnectedStatus` in the UI as a single type of filter control?

Like, would API users ever do something like:

```
/service-groups?filter=status:critical&filter=disconnectedStatus:disconnected
```

and expect to be able to do the same filtering via the UI?

**Comment by apriofrost** [comment]
hi @scottopherson , this is a great critique! 
The UI solution presented here is just a stopgap before we get to the point where the status filters here are multi-selectable. 

There are two types of filters in the service groups page:
### Connectivity: connected and disconnected
Users can filter out the disconnected ones to decide if they want to remove them from the view or troubleshooting what went wrong.
- For services, "disconnected" means the last health check report for this service was received X time ago while X is larger than the threshold users set (default: 5 minutes)
- For service groups, disconnected means whether or not a service group contains _any_ services that are marked as disconnected
### Health status: critical, warning, unknown and OK
Users can filter out the critical, warning and unknown ones to troubleshoot; We have plans to remove the OK filter in the future and have the OK count live outside the filters just as informational.
- For services, the health status matches with the output from the health check hook from Habitat supervisors.
- For service groups, the health status is roll up from the services. Currently, one group can only have one health status: any service groups contain any critical services are marked as critical; if not, any that contains unknown is marked as unknown; if not, any that contains warning n is marked as warning; else, marked as OK. HOWEVER, We have plans to simplify that logic to remove the construct of group health status so that the `filter= status: critical` means that any service groups that contain any services in a critical state (matches with the "disconnected" filtering logic)

### TBD: Other filters
In the future, we have plans to add other types of filters, such as filtering on deployment status: deploying; failed deployment, etc.

## Vision for multi-selectable filters
Here is a picture showing how the multi-selectable status filters could look like.
In this case, service groups that contain any services that are either critical, warning, or disconnected are filtered out in the table (the logic between all filters are OR). There is another idea that we are considering is grouping the filters by type, and make the logic between different types AND, and same types OR. It would make more sense when we add more types of filters in the future, such as the deployment status.
![image](https://user-images.githubusercontent.com/9401459/66879480-1be85800-ef73-11e9-933b-f41af7862f51.png)

In short, I agree with a lot about what you suggested above. We could have a chat with @kmacgugan about how we want to redesign the API going forward to support the long term vision.

**Comment by kmacgugan** [comment]
We separated the naming of the filter on the sidebar to be health and the filter on the main table to be status so we can tease them apart from a URL. When we had them all under the same label we were unable to distinguish which filters were applied to the table and which to apply to the services side bar. 

I think the sidebar filter of ```filter=disconnectedStatus:disconnected``` might be erroneous and we might want to make it ```health=disconnected```.

I think we could also make the main table be 
```filter=disconnected```

It's sort of splitting a semantic hair if you consider disconnected to be part of the status/health.

**Comment by apriofrost** [comment]
@kmacgugan agree with having separate naming for the filters on the sidebar for services in a group, and the filter on the main table for service groups.

I would suggest making "disconnected" a separate type from the health. Because a service has both a connectivity status and a health status. For example, a service's last health check report returned "OK". But it was received 10 minutes ago (passing the 5 minutes threshold), so it's also marked as "disconnected".

Something like this could possibly happen:
```
filter=health: critical
filter=health: warning
filter=health: unknown
filter=connectivity: disconnected
filter=deployment: failed
filter=deployment: pending
filter=last-deployed: 30min
filter=last-deployed: 1hr
(and others, we plan to get rid of "total" and "OK)
```

For different types of filters (`health`, `connectivity`, `deployment`, and `last-deployed`), the filter logic is `and`; Within each type, the logic is `or`.

So that a user could filter: show me all the service groups that are last deployed 30 minutes ago, which has
failed, and the health status is critical or warning.

The query could look like this:

```
/service-groups?filter=last-deployed:30min&filter=deployment:failed&filter=health:critical, warning
```

(note: just an example, I'm not sure if `,` is the right way to indicate `or`)

**Comment by apriofrost** [comment]
This is a related issue we found out about the URL inconsistency (https://github.com/chef/automate/issues/1884) , maybe it can be addressed within this task?  @scottopherson

**Comment by danielsdeleo** [comment]
Kyleen changed the API to make disconnected appear as a single status of the same form as the health statuses in #2071
